frb_dart/lib/src/ffi/io.dart

```diff
@@ -42,35 +29,177 @@ class DartApiDl {
+  static _TypedData _typedDataIntoDart(
+    int ty,
+    ffi.Pointer<ffi.Uint8> typedValues,
+    int nValues,
+  ) {
+    switch (ty) {
+      case DartTypedDataType.ByteData:
+        return _TypedData<ByteData>(
+          ByteData.view(
+            typedValues.cast<ffi.Uint8>().asTypedList(nValues).buffer,
+          ),
+          (view) => ByteData.view(
+            Uint8List.fromList(view.buffer.asUint8List()).buffer,
+          ),
+        );
+      case DartTypedDataType.Int8:
+        final view = typedValues.cast<ffi.Int8>().asTypedList(nValues);
+        return _TypedData<Int8List>(view, Int8List.fromList);
+      case DartTypedDataType.Uint8:
+        final view = typedValues.cast<ffi.Uint8>().asTypedList(nValues);
+        return _TypedData<Uint8List>(view, Uint8List.fromList);
+      case DartTypedDataType.Int16:
+        final view = typedValues.cast<ffi.Int16>().asTypedList(nValues);
+        return _TypedData<Int16List>(view, Int16List.fromList);
+      case DartTypedDataType.Uint16:
+        final view = typedValues.cast<ffi.Uint16>().asTypedList(nValues);
+        return _TypedData<Uint16List>(view, Uint16List.fromList);
+      case DartTypedDataType.Int32:
+        final view = typedValues.cast<ffi.Int32>().asTypedList(nValues);
+        return _TypedData<Int32List>(view, Int32List.fromList);
+      case DartTypedDataType.Uint32:
+        final view = typedValues.cast<ffi.Uint32>().asTypedList(nValues);
+        return _TypedData<Uint32List>(view, Uint32List.fromList);
+      case DartTypedDataType.Int64:
+        final view = typedValues.cast<ffi.Int64>().asTypedList(nValues);
+        return _TypedData<Int64List>(view, Int64List.fromList);
+      case DartTypedDataType.Uint64:
+        final view = typedValues.cast<ffi.Uint64>().asTypedList(nValues);
+        return _TypedData<Uint64List>(view, Uint64List.fromList);
+      case DartTypedDataType.Float32:
+        final view = typedValues.cast<ffi.Float>().asTypedList(nValues);
+        return _TypedData<Float32List>(view, Float32List.fromList);
+      case DartTypedDataType.Float64:
+        final view = typedValues.cast<ffi.Double>().asTypedList(nValues);
+        return _TypedData<Float64List>(view, Float64List.fromList);
 
-  Uint8List? get buffer {
-    if (ptr.address == 0) {
-      return null;
+      case DartTypedDataType.Uint8Clamped:
+      case DartTypedDataType.Float32x4:
+      case DartTypedDataType.Invalid:
+      default:
+        throw Exception("Can't read invalid typed data type $ty");
     }
-    return Uint8List.fromList(ptr.asTypedList(len));
   }
 
-  bool get isSuccess => success > 0;
+  static final _externalTypedDataFinalizer =
+      Finalizer<_ExternalTypedDataFinalizerArgs>((externalTypedData) {
+    final handleFinalizer =
+        externalTypedData.callback.asFunction<DartExternalTypedDataFinalizer>();
+    handleFinalizer(externalTypedData.length, externalTypedData.peer);
+
+    if (Platform.environment.containsKey('FRB_TEST')) {
```

Comment:
uber-nit: if we want best performance, would be great to use https://api.flutter.dev/flutter/dart-core/bool/bool.fromEnvironment.html which is compile time constant (instead of the Platform.environment which is runtime "environment variable"). By doing so, if someone do not define FRB_TEST, then this `if` will be removed at compile time so zero runtime cost.

p.s. defining that constant needs `--dart-define` instead of normal env var

Reply:
Looks like you gave the missing piece of my puzzle, thanks :)

Reply:
You are welcome!